lib/create_azure_oauth_config.ts

```diff
@@ -0,0 +1,97 @@
+ * Returns the OAuth configuration for Azure.
+ *
+ * Requires `--allow-env[=AZURE_CLIENT_ID,AZURE_CLIENT_SECRET,AZURE_CLOUD_INSTANCE,AZURE_TENANT_ID]`
+ * permissions and environment variables:
+ * 1. `AZURE_CLIENT_ID`
+ * 2. `AZURE_CLIENT_SECRET`
+ * 3. `AZURE_CLOUD_INSTANCE`
+ * 4. `AZURE_TENANT_ID`
+ *
+ * @example
+ * ```ts
+ * import { createAzureOAuthConfig } from "https://deno.land/x/deno_kv_oauth@$VERSION/mod.ts";
+ *
+ * const oauthConfig = createAzureOAuthConfig({
+ *   envPrefix: "AZURE",
+ *   redirectUri: "http://localhost:8000/callback",
+ *   scope: "openid",
+ * });
+ * ```
+ *
+ * You can use a separate environment variable prefix with the envPrefix configureation.
+ * Requires `--allow-env[=MY_PREFIX_CLIENT_ID,MY_PREFIX_CLIENT_SECRET,MY_PREFIX_CLOUD_INSTANCE,MY_PREFIX_TENANT_ID]`
+ * permissions and environment variables:
+ * 1. `MY_PREFIX_CLIENT_ID`
+ * 2. `MY_PREFIX_CLIENT_SECRET`
+ * 3. `MY_PREFIX_CLOUD_INSTANCE`
+ * 4. `MY_PREFIX_TENANT_ID`
+ *
+ * @example
+ * ```ts
+ * import { createAzureOAuthConfig } from "https://deno.land/x/deno_kv_oauth@$VERSION/mod.ts";
+ *
+ * const oauthConfig = createAzureOAuthConfig({
+ *   envPrefix: "MY_PREFIX",
+ *   redirectUri: "http://localhost:8000/callback",
+ *   scope: "openid",
+ * });
+ * ```
+ *
+ * This also supports the ability to connect with Azure AD B2C instances using the
+ * proper configurations for AD B2C instances
+ *
+ * @see {@link https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc}
+ */
+export function createAzureOAuthConfig(config: {
+  /** Used to set the prefix used to access environment variables; defaults to `AZURE`.*/
+  envPrefix?: string;
+  /** @see {@linkcode OAuth2ClientConfig.redirectUri} */
+  redirectUri?: string;
+  /** @see {@linkcode OAuth2ClientConfig.defaults.scope} */
+  scope: string | string[];
+}): OAuth2ClientConfig {
+  const getEnvKey = (key: string) => {
+    return `${config.envPrefix || "AZURE"}_${key}`;
+  };
+
+  let cloudInstance = getRequiredEnv(getEnvKey("CLOUD_INSTANCE"));
+
+  if (!cloudInstance.endsWith("/")) {
+    cloudInstance = `${cloudInstance}/`;
+  }
+
+  const tenantId = getRequiredEnv(getEnvKey("TENANT_ID"));
+
+  const policy = Deno.env.get(getEnvKey("POLICY"));
+
+  const path = policy ? `${tenantId}/${policy}` : tenantId;
+
+  const baseURL = `${cloudInstance}${path}/oauth2`;
+
+  const clientId = getRequiredEnv(getEnvKey("CLIENT_ID"));
+
+  if (Array.isArray(config.scope) && config.scope.some((s) => s === "openid")) {
+    config.scope.push(clientId);
+  } else if (
+    typeof config.scope === "string" &&
+    config.scope.includes("openid")
+  ) {
+    config.scope = `${config.scope} ${clientId}`;
+  }
+
+  return {
+    clientId,
+    clientSecret: getRequiredEnv(getEnvKey("CLIENT_SECRET")),
+    authorizationEndpointUri: `${baseURL}/v2.0/authorize`,
+    tokenUri: `${baseURL}/v2.0/token`,
+    redirectUri: config.redirectUri,
+    defaults: {
+      scope: config.scope,
+    },
+  };
+}
```

Comment:
```suggestion
/**
 * Returns the OAuth configuration for Azure.
 *
 * Requires `--allow-env[=AZURE_CLIENT_ID,AZURE_CLIENT_SECRET,AZURE_TENANT_ID]`
 * permissions and environment variables:
 * 1. `AZURE_CLIENT_ID`
 * 2. `AZURE_CLIENT_SECRET`
 * 3. `AZURE_TENANT_ID`
 *
 * @example
 * ```ts
 * import { createAzureOAuthConfig } from "https://deno.land/x/deno_kv_oauth@$VERSION/mod.ts";
 *
 * const oauthConfig = createAzureOAuthConfig({
 *   redirectUri: "http://localhost:8000/callback",
 *   scope: "https://graph.microsoft.com/User.Read",
 * });
 * ```
 *
 * @see {@link https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-auth-code-flow}
 */
export function createAzureOAuthConfig(config: {
  /** @see {@linkcode OAuth2ClientConfig.redirectUri} */
  redirectUri?: string;
  /** @see {@linkcode OAuth2ClientConfig.defaults.scope} */
  scope: string | string[];
}): OAuth2ClientConfig {
  const baseUrl = `https://login.microsoftonline.com/${
    getRequiredEnv("AZURE_TENANT_ID")
  }/oauth2/v2.0`;
  return {
    clientId: getRequiredEnv("AZURE_CLIENT_ID"),
    clientSecret: getRequiredEnv("AZURE_CLIENT_SECRET"),
    authorizationEndpointUri: `${baseUrl}/authorize`,
    tokenUri: `${baseUrl}/token`,
    redirectUri: config.redirectUri,
    defaults: {
      scope: config.scope,
    },
  };
}
```
These changes bring the Azure implementation into line with the other platforms. I've tested it, and it works as expected. I want to clarify that these functions are meant to cater to the majority of use cases, not all use cases. Advanced use cases can take advantage of custom OAuth configs, if needed.

Reply:
With these changes azure ADB2C won't work, which is actually what we need to use, the major issue being the hardcoding of https://login.microsoftonline.com.  I split AD and ADB2C into separate config providers. Though i still think we should include the logic around client ID, otherwise the end user experience is more complicated by having to know to add the client ID as a scope instead of just `openid`.

Reply:
Let me know if you want me to remove the client ID logic.

Tests were added as well.